src/backend/distributed/shared_library_init.c

```diff
@@ -927,16 +949,38 @@ RegisterCitusConfigVariables(void)
 		GUC_STANDARD,
 		NULL, NULL, NULL);
 
+	DefineCustomIntVariable(
+		"citus.max_shared_pool_size",
+		gettext_noop("Sets the maximum number of connections allowed per worker node "
+					 "across all the backends from this node. Setting to -1 disables "
+					 "connections throttling. Setting to 0 makes it auto-adjust, meaning "
+					 "equal to max_connections."),
+		gettext_noop("For single coordinator setups, the value should be adjusted "
+					 "to match the max_connections on the remote nodes. For Citus MX, "
+					 "the value should be tuned such that it is roughly "
+					 "(max_connections / #workerNodes) so that each worker doesn't "
+					 "become overwhelmed by too many incoming connections"),
+		&MaxSharedPoolSize,
+		0, -1, INT_MAX,
+		PGC_SIGHUP, /* TODO: is PGC_POSTMASTER more convinient? */
```

Comment:
I'm split on  `PGC_SIGHUP ` vs `PGC_POSTMASTER `.

For emergency situations, where we want to set `MaxSharedPoolSize` to `-1`, `PGC_SIGHUP` sounds pretty reasonable.

But, in cases where `MaxSharedPoolSize` is changed such that it is lower than the active connection count, the counters might become above the limit.

Anyway, I'll mark this `GUC_SUPERUSER_ONLY` as this shouldn't be done by regular users.

Any comments on this?

Reply:
I think SIGHUP makes sense. Is it bad for counters to go above the limit?

Reply:
>  Is it bad for counters to go above the limit?

I don't think there are any issues that'd break anything or such. But, that's something not intended in the code. Let me run some benchmarks, and concurrently changing `MaxSharedPoolSize ` :)

Reply:
My observation/tests on this:

When a user decreases `MaxSharedPoolSize`, and  we already have more cached connections than `MaxSharedPoolSize`, no new backend will ever get a chance to get a connection slot. Plus, the counter would never go down as long as the cached  connections stay.  I think that is expected.

Other than that, all seems to work as expected.